97668326 whole building load (Issue# 4984)

rongpengzhang opened the pull request:
The time-step whole building heating/cooling load can be reported by setting up two meters, i.e., HeatingCoils:EnergyTransfer and CoolingCoils:EnergyTransfer, respectively. In the current version EnergyPlus, 20 types of heating equipment are included in the HeatingCoils:EnergyTransfer calculation and 15 types of cooling equipment are included in the CoolingCoils:EnergyTransfer calculation. However, some equipment may be missing.

In this branch, the following equipment are added:
- Coil:WaterHeating:AirToWaterHeatPump:Pumped 
- Coil:WaterHeating:AirToWaterHeatPump:Wrapped
  -> The above two equipment were added in the CoolingCoils:EnergyTransfer meter if the evaporator is placed indoors; if outdoors, they are not included.
- ZoneHVAC:EvaporativeCoolerUnit
- Coil:UserDefined
  -> update EMS src to handle HeatingCoils:EnergyTransfer and
  CoolingCoils:EnergyTransfer  meter
  -> add the appropriate EnergyManagementSystem:MeteredOutputVariable
  objects to the userdefined example files
  -> mention this in the documentation
- Coil:WaterHeating:AirToWaterHeatPump:VariableSpeed 
  This equipment was included in the meter in the exsiting codes. But it did not distinguish the indoor  evaporator and outdoor evaporator, so the src were modified.

rongpengzhang committed:
Whole Building Load V1

rongpengzhang committed:
Merge branch 'develop' into 97668326-WholeBuildingLoad

rongpengzhang committed:
Whole Bldg Load: V1-3

V1:  Add Coil:WaterHeating:AirToWaterHeatPump:Wrapped &
Coil:WaterHeating:AirToWaterHeatPump:Pumped to the
CoolingCoils:EnergyTransfer meter.

V2:  Modify Coil:WaterHeating:AirToWaterHeatPump:VariableSpeed

V3: Add ZoneHVAC:EvaporativeCoolerUnit to the
CoolingCoils:EnergyTransfer meter.

rongpengzhang committed:
Whole Bldg Load: V4

Add Coil:UserDefined.
>> update EMS src to handle HeatingCoils:EnergyTransfer and
CoolingCoils:EnergyTransfer  meter
>> add the appropriate EnergyManagementSystem:MeteredOutputVariable
objects to the userdefined example files
>> mention this in the documentation and

rongpengzhang commented:
Hi @mjwitte, I have finished this issue as what we discussed. Can you please have a review?

mjwitte commented:
@rongpengzhang Are there any test idf files or a modified example file to illustrate use of these meters and confirm that they line up with other outputs?

rongpengzhang commented:
@mjwitte The tested file can be found via https://www.dropbox.com/sh/ul41dev8fx6g9fh/AABMLHN-Su57k47gzj9LO-gEa?dl=0

"~Test Result_0913.docx" is a summary of the test results.

mjwitte commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -2957,12 +2958,22 @@ namespace DXCoils {
 
+			//Check if the air inlet node is OA node, to justify whether the coil is placed in zone or not
+			EnergyPlus::OutAirNodeManager::CheckAndAddAirNodeNumber( DXCoil( DXCoilNum ).AirInNode, IsOANodeListed );
+			if ( IsOANodeListed ) {
```
@rongpengzhang This is not the right function to use.  This will check if the node is on an OA node list, but it will also add the node to the OA list, causing it to be initialized with outdoor air conditions. Use CheckOutAirNodeNumber instead.  Also, preferred style is to add "using OutAirNodeManager::CheckOutAirNodeNumber" rather than reference the function with full namespaces.

rongpengzhang replied:
Thanks Mike. I will use CheckOutAirNodeNumber following the preferred style.

mjwitte commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -3381,11 +3392,20 @@ namespace DXCoils {
 			TestCompSet( CurrentModuleObject, Alphas( 1 ), Alphas( 3 ), Alphas( 4 ), "Air Nodes" );
+			
+			//Check if the air inlet node is OA node, to justify whether the coil is placed in zone or not
+			EnergyPlus::OutAirNodeManager::CheckAndAddAirNodeNumber( DXCoil( DXCoilNum ).AirInNode, IsOANodeListed );
```
@rongpengzhang Same comments.

rongpengzhang replied:
Fixed.

mjwitte commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -1963,12 +1956,22 @@ namespace VariableSpeedCoils {
 			TestCompSet(CurrentModuleObject, AlphArray(1), AlphArray(5), AlphArray(6), "Air Nodes");
 
+			//Check if the air inlet node is OA node, to justify whether the coil is placed in zone or not
+			EnergyPlus::OutAirNodeManager::CheckAndAddAirNodeNumber( VarSpeedCoil( DXCoilNum ).AirInletNodeNum, IsOANodeListed );
```
@rongpengzhang And here also.

rongpengzhang replied:
Fixed.

mjwitte commented:
Unit test?

rongpengzhang commented:
@mjwitte Do we need unit tests for this issue? I didn't do that since there is no new added methods. But I did tests with example files locally.

rongpengzhang committed:
Modify method to check coil location

- Use CheckOutAirNodeNumber method to replace CheckAndAddAirNodeNumber
method
- Develop unit test for this

rongpengzhang committed:
Merge branch 'develop' into 97668326-WholeBuildingLoad

Conflicts:
	tst/EnergyPlus/unit/DXCoils.unit.cc

rongpengzhang committed:
Merge conflicts solved

Myoldmopar commented:
> Do we need unit tests for this issue?

Yes

Re-assigning this to 8.4.0+BugFix milestone.

mjwitte commented:
@Myoldmopar Unit test was added yesterday, so this could still go in for v8.4.0

mjwitte commented:
But there are some build warnings.

rongpengzhang commented:
@mjwitte Build warnings fixed.

rongpengzhang committed:
Build warnings fixed

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 97668326-WholeBuildingLoad
